src/software/ai/hl/stp/tactic/penalty_goalie_tactic.cpp

```diff
@@ -0,0 +1,298 @@
+
+#include "software/ai/hl/stp/tactic/penalty_goalie_tactic.h"
+
+#include "shared/constants.h"
+#include "software/ai/evaluation/calc_best_shot.h"
+#include "software/geom/algorithms/calculate_block_cone.h"
+#include "software/geom/algorithms/closest_point.h"
+#include "software/geom/algorithms/contains.h"
+#include "software/geom/algorithms/intersection.h"
+#include "software/geom/line.h"
+#include "software/geom/point.h"
+#include "software/geom/ray.h"
+#include "software/geom/segment.h"
+
+PenaltyGoalieTactic::PenaltyGoalieTactic(const Ball &ball, const Field &field,
+                           const Team &friendly_team, const Team &enemy_team,
+                           std::shared_ptr<const GoalieTacticConfig> goalie_tactic_config)
+    : Tactic(true, {RobotCapability::Move}),
+      ball(ball),
+      field(field),
+      friendly_team(friendly_team),
+      enemy_team(enemy_team),
+      goalie_tactic_config(goalie_tactic_config)
+{
+}
+
+Point PenaltyGoalieTactic::restrainPenaltyGoalieInGoalLine(Point goalie_desired_position)
+{
+    double upper_y_goal_line = field.friendlyGoalCenter().y() + field.goalYLength() / 2;
+    double lower_y_goal_line = field.friendlyGoalCenter().y() - field.goalYLength() / 2;
+
+    Point upper_point_goal_line = Point(field.friendlyGoalCenter().x(), upper_y_goal_line);
+
+    Point lower_point_goal_line = Point(field.friendlyGoalCenter().x(), lower_y_goal_line);
+
+    Segment goal_line = Segment(upper_point_goal_line, lower_point_goal_line);
+
+    // if the goalie restricted area already contains the point, then we are
+    // safe to move there.
+    if (contains(goal_line, goalie_desired_position))
+    {
+        return Point(goalie_desired_position);
+    } else {
+        Line first = Line(goalie_desired_position, field.friendlyGoalCenter());
+        Line second = Line(upper_point_goal_line, lower_point_goal_line);
+
+        // calculate intersection
+        double a1          = first.getCoeffs().a;
+        double b1          = first.getCoeffs().b;
+        double c1          = first.getCoeffs().c;
+        double a2          = second.getCoeffs().a;
+        double b2          = second.getCoeffs().b;
+        double c2          = second.getCoeffs().c;
+        double determinant = (a1 * b2) - (a2 * b1);
+
+        double x = ((b1 * c2) - (b2 * c1)) / determinant;
+        double y = ((a2 * c1) - (a1 * c2)) / determinant;
```

Comment:
seems like we should have a similar function in geom

Reply:
I initially did this to get around having to use an optional pointer but I refactored it now to use intersection from geom